rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/event/RMInitialState.java

```diff
@@ -163,41 +147,42 @@ private void updateCounter(RMEvent event) {
+    public RMStateDTO cloneAndFilter(long filter) {
+        long actualFilter = computeActualFilter(filter);
+
+        final List<RMEvent> responseEvents = events.getSortedItems()
+                                                   .tailSet(new RMEvent(actualFilter + 1)) // because tailSet returns event which is equal or greater
+                                                   .stream()
+                                                   .limit(PAResourceManagerProperties.RM_REST_MONITORING_MAXIMUM_CHUNK_SIZE.getValueAsInt())
+                                                   .collect(Collectors.toList());
+
+        RMStateDTO response = new RMStateDTO();
+
+        response.setNodeSource(getNodeSource(responseEvents));
+        response.setNodesEvents(getNodesEvents(responseEvents));
+        response.setLatestCounter(Math.max(actualFilter, findLargestCounter(responseEvents)));
+
+        return response;
+    }
+
+    private long computeActualFilter(long clientFilter) {
+        if (clientFilter <= latestCounter.get()) {
+            return clientFilter;
         } else {
             LOGGER.info(String.format("Client is aware of %d but server knows only about %d counter. " +
                                       "Probably because there was network server restart.",
-                                      filter,
+                                      clientFilter,
                                       latestCounter.get()));
-            actualFilter = -1; // reset filter to default  value
+            return EMPTY_STATE; // reset filter to default  value
         }
-        RMInitialState clone = new RMInitialState();
-
-        clone.nodeEvents = newFilteredEvents(this.nodeEvents, actualFilter);
-        clone.nodeSourceEvents = newFilteredEvents(this.nodeSourceEvents, actualFilter);
-
-        clone.latestCounter.set(Math.max(actualFilter,
-                                         Math.max(findLargestCounter(clone.nodeEvents.values()),
-                                                  findLargestCounter(clone.nodeSourceEvents.values()))));
-        return clone;
     }
 
-    private <T extends RMEvent> Map<String, T> newFilteredEvents(Map<String, T> events, long filter) {
-        return events.entrySet()
-                     .stream()
-                     .filter(entry -> entry.getValue().getCounter() > filter)
-                     .collect(Collectors.toMap(entry -> entry.getKey(), entry -> entry.getValue()));
+    private void updateCounter(RMEvent event) {
+        latestCounter.set(Math.max(latestCounter.get(), event.getCounter()));
     }
 
     private <T extends RMEvent> long findLargestCounter(Collection<T> events) {
         final Optional<T> max = events.stream().max(Comparator.comparing(RMEvent::getCounter));
-        if (max.isPresent()) {
-            return max.get().getCounter();
-        } else {
-            return 0;
-        }
+        return max.map(RMEvent::getCounter).orElse(0l);
```

Comment:
What is the case `0l` about? Can we define a constant for it? Can we use `EMPTY_STATE` instead (-1)?

Reply:
Yes, right